File: iosApp/iosApp/Features/Journal/Loggers.swift
import SwiftUI
import Shared

// Shared vocabulary for the mood scale.
enum MoodScale {
    static let emojis = ["😫", "😕", "😐", "🙂", "😄"]
    static let labels = ["Rough", "Meh", "Mid", "Good", "Lit"]
    static func emoji(_ v: Int) -> String { emojis[min(max(v - 1, 0), 4)] }
    static func label(_ v: Int) -> String { labels[min(max(v - 1, 0), 4)] }
    /// Diverging valence ramp: deep red fades to a NEUTRAL midpoint, then greens deepen.
    /// The old scale reused brand colours (rose for "meh" read as positive, and "good"/"lit"
    /// were two near-identical greens). These five are validated for adjacent-step
    /// distinguishability, incl. colour-blind simulation, against the cream card surface.
    /// Colour is never the only carrier: the legend labels each dot, charts encode mood by
    /// position, and the sheets pair every value with its emoji.
    static func color(_ v: Int) -> Color {
        switch v {
        case 1:  return Color(hex: "B03A3A")   // rough — deep red
        case 2:  return Color(hex: "EC9A82")   // meh — washed salmon
        case 3:  return Color(hex: "847A6B")   // mid — neutral warm grey (the midpoint)
        case 4:  return Color(hex: "6DBA7F")   // good — light green
        default: return Color(hex: "2E7B50")   // lit — deep green
        }
    }
}

enum HeadacheScale {
    static let labels = ["", "Mild", "Moderate", "Severe"]
    static func label(_ v: Int) -> String { labels[min(max(v, 0), 3)] }
}

/// Log or edit the mood for a given day.
struct MoodSheet: View {
    @Environment(CycleStore.self) private var store
    @Environment(\.dismiss) private var dismiss
    let dateISO: String

    @State private var vibe: Int?
    @State private var note = ""

    var body: some View {
        NavigationStack {
            Form {
                Section("How's the vibe?") {
                    HStack(spacing: 6) {
                        ForEach(1...5, id: \.self) { v in
                            Button {
                                withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) { vibe = v }
                            } label: {
                                VStack(spacing: 4) {
                                    Text(MoodScale.emoji(v))
                                        .font(.system(size: 28))
                                        .opacity(vibe == nil || vibe == v ? 1 : 0.4)
                                    Text(MoodScale.label(v))
                                        .font(.caption2)
                                        .foregroundStyle(vibe == v ? Theme.primary : Theme.inkSoft)
                                }
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 6)
                                .background(vibe == v ? Theme.primary.opacity(0.1) : .clear,
                                            in: RoundedRectangle(cornerRadius: 12))
                            }
                            .buttonStyle(.plain)
                        }
                    }
                }
                Section("Note (optional)") {
                    TextField("What's on your mind?", text: $note, axis: .vertical).lineLimit(2...5)
                }
                if store.mood(on: dateISO) != nil {
                    Button("Remove mood", role: .destructive) { store.clearMood(on: dateISO); dismiss() }
                }
            }
            .navigationTitle("Mood").navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Cancel") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Save") {
                        if let v = vibe {
                            store.logMood(v, note: note.trimmingCharacters(in: .whitespacesAndNewlines), on: dateISO)
                        }
                        dismiss()
                    }
                    .disabled(vibe == nil)
                }
            }
            .onAppear {
                if let m = store.mood(on: dateISO) { vibe = Int(m.mood); note = m.note }
            }
        }
    }
}

/// Log or edit one headache / migraine episode. Multiple per day are supported, so this
/// always adds a new episode unless editing an [existing] one.
struct HeadacheSheet: View {
    @Environment(CycleStore.self) private var store
    @Environment(\.dismiss) private var dismiss
    let dateISO: String
    var existing: HeadacheLog?

    @State private var intensity = 2
    @State private var time = Date()
    @State private var symptoms: Set<String> = []
    @State private var triggers: Set<String> = []
    @State private var note = ""

    private let symptomOptions = ["Throbbing", "One-sided", "Nausea", "Light sensitivity",
                                  "Sound sensitivity", "Visual aura", "Tingling"]
    private let triggerOptions = ["Poor sleep", "Skipped meal", "Stress", "Dehydration",
                                  "Caffeine", "Alcohol", "Weather", "Strong smells", "Hormones"]

    var body: some View {
        NavigationStack {
            Form {
                Section("Intensity") {
                    Picker("Intensity", selection: $intensity) {
                        Text("Mild").tag(1); Text("Moderate").tag(2); Text("Severe").tag(3)
                    }
                    .pickerStyle(.segmented)
                }
                Section("Time") {
                    DatePicker("Onset", selection: $time, displayedComponents: .hourAndMinute)
                }
                Section("Symptoms") { ChipWrap(options: symptomOptions, selection: $symptoms) }
                Section("Possible triggers") { ChipWrap(options: triggerOptions, selection: $triggers) }
                Section("Note (optional)") {
                    TextField("Anything to remember", text: $note, axis: .vertical).lineLimit(2...4)
                }
                if existing != nil {
                    Button("Delete this episode", role: .destructive) {
                        store.deleteHeadache(id: existing!.id); dismiss()
                    }
                }
            }
            .navigationTitle(existing == nil ? "Headache" : "Edit headache")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Cancel") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) { Button("Save") { save() } }
            }
            .onAppear(perform: prefill)
        }
    }

    private func save() {
        let at = combinedDate()
        let n = note.trimmingCharacters(in: .whitespacesAndNewlines)
        if let e = existing {
            store.updateHeadache(id: e.id, intensity: intensity, symptoms: Array(symptoms),
                                 triggers: Array(triggers), note: n, at: at)
        } else {
            store.addHeadache(intensity: intensity, symptoms: Array(symptoms),
                              triggers: Array(triggers), note: n, at: at)
        }
        dismiss()
    }

    /// Combine the browsed day with the chosen time.
    private func combinedDate() -> Date {
        let cal = Calendar.current
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"; f.timeZone = .current
        let day = f.date(from: dateISO) ?? Date()
        let tc = cal.dateComponents([.hour, .minute], from: time)
        return cal.date(bySettingHour: tc.hour ?? 9, minute: tc.minute ?? 0, second: 0, of: day) ?? day
    }

    private func prefill() {
        guard let e = existing else { return }
        intensity = Int(e.intensity)
        symptoms = Set(e.symptoms)
        triggers = Set(e.triggers)
        note = e.note
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd'T'HH:mm"; f.timeZone = .current
        if let d = f.date(from: e.at) { time = d }
    }
}

/// A wrapping set of multi-select chips.
struct ChipWrap: View {
    let options: [String]
    @Binding var selection: Set<String>

    var body: some View {
        LazyVGrid(columns: [GridItem(.adaptive(minimum: 104), spacing: 8)], alignment: .leading, spacing: 8) {
            ForEach(options, id: \.self) { opt in
                let on = selection.contains(opt)
                Button {
                    if on { selection.remove(opt) } else { selection.insert(opt) }
                } label: {
                    Text(opt)
                        .font(.footnote)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 7).padding(.horizontal, 8)
                        .foregroundStyle(on ? .white : Theme.ink)
                        .background(on ? AnyShapeStyle(Theme.primary) : AnyShapeStyle(Theme.background),
                                    in: Capsule())
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.vertical, 4)
    }
}

/// "Headaches & your cycle" — the hormonal-cluster insight, only shown when confident.
struct HeadacheInsightCard: View {
    let insight: HeadacheInsight

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Image(systemName: "bolt.heart.fill").foregroundStyle(Theme.secondary)
            VStack(alignment: .leading, spacing: 4) {
                Text("Headaches & your cycle").font(.cyclunaSerif(19)).foregroundStyle(Theme.ink)
                Text("Your headaches tend to cluster in your \(phaseWord(insight.phase)) — a common hormonal pattern (\(Int(insight.count)) of \(Int(insight.total)) so far).")
                    .font(.footnote).foregroundStyle(Theme.inkSoft)
                    .fixedSize(horizontal: false, vertical: true)
            }
            Spacer(minLength: 0)
        }
        .cyclunaCard()
    }

    private func phaseWord(_ phase: Phase) -> String {
        switch phase.label {
        case "Ovulatory":  return "ovulation window"
        case "Follicular": return "follicular phase"
        case "Luteal":     return "luteal phase"
        default:           return "period"
        }
    }
}

/// Write a new journal note, or edit an existing one, for a given day.
struct NoteSheet: View {
    @Environment(CycleStore.self) private var store
    @Environment(\.dismiss) private var dismiss
    let dateISO: String
    var existing: JournalEntry?

    @State private var text = ""

    /// UI-level cap only. The core stays permissive so an imported entry longer than this
    /// is never silently truncated on the way in.
    private static let maxLength = 2_000

    private var remaining: Int { Self.maxLength - text.count }

    var body: some View {
        NavigationStack {
            VStack(alignment: .trailing, spacing: 4) {
                TextEditor(text: $text)
                    .font(.system(.body, design: .serif))
                    .onChange(of: text) { _, new in
                        if new.count > Self.maxLength { text = String(new.prefix(Self.maxLength)) }
                    }
                // Only surface the limit as it comes into view — a counter on an empty note
                // is just clutter.
                if remaining <= 200 {
                    Text("\(remaining) characters left")
                        .font(.caption2)
                        .foregroundStyle(remaining <= 0 ? Theme.phaseMenstrual : Theme.inkSoft)
                }
            }
            .padding()
                .navigationTitle(existing == nil ? "New note" : "Edit note")
                .navigationBarTitleDisplayMode(.inline)
                .toolbar {
                    ToolbarItem(placement: .cancellationAction) { Button("Cancel") { dismiss() } }
                    ToolbarItem(placement: .confirmationAction) {
                        Button("Save") {
                            let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
                            if !t.isEmpty {
                                if let e = existing {
                                    store.updateJournalEntry(id: e.id, text: t, on: dateISO)
                                } else {
                                    store.addJournalEntry(text: t, on: dateISO)
                                }
                            }
                            dismiss()
                        }
                    }
                    if let e = existing {
                        ToolbarItem(placement: .bottomBar) {
                            Button("Delete", role: .destructive) { store.deleteJournalEntry(id: e.id); dismiss() }
                        }
                    }
                }
                .onAppear { if let e = existing { text = e.text } }
        }
    }
}

File: androidApp/src/main/kotlin/app/cycluna/android/features/journal/Loggers.kt
package app.cycluna.android.features.journal

import android.text.format.DateFormat
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.TimePicker
import androidx.compose.material3.TimePickerDefaults
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.material3.rememberTimePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.toMutableStateList
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cycluna.android.LocalCycleStore
import app.cycluna.android.core.HeadacheScale
import app.cycluna.android.core.MoodScale
import app.cycluna.android.designsystem.HeartGlyph
import app.cycluna.android.designsystem.Theme
import app.cycluna.android.designsystem.cyclunaCard
import app.cycluna.android.designsystem.serif
import app.cycluna.core.HeadacheLog
import app.cycluna.core.JournalEntry
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

private val SYMPTOM_OPTIONS = listOf(
    "Throbbing", "One-sided", "Nausea", "Light sensitivity",
    "Sound sensitivity", "Visual aura", "Tingling",
)
private val TRIGGER_OPTIONS = listOf(
    "Poor sleep", "Skipped meal", "Stress", "Dehydration",
    "Caffeine", "Alcohol", "Weather", "Strong smells", "Hormones",
)

/** UI-level cap only. The core stays permissive so an imported entry is never truncated. */
private const val MAX_NOTE_LENGTH = 2_000

// ------------------------------------------------------------------------------------
// Mood
// ------------------------------------------------------------------------------------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MoodSheet(dateIso: String, onDismiss: () -> Unit) {
    val store = LocalCycleStore.current
    val existing = store.mood(dateIso)
    var vibe by remember { mutableStateOf(existing?.mood) }
    var note by remember { mutableStateOf(existing?.note ?: "") }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = Theme.surface,
    ) {
        SheetFrame(
            title = "Mood",
            saveEnabled = vibe != null,
            onCancel = onDismiss,
            onSave = {
                vibe?.let { store.logMood(it, note.trim(), dateIso) }
                onDismiss()
            },
        ) {
            SectionLabel("How's the vibe?")
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                (1..5).forEach { v ->
                    val chosen = vibe == v
                    val scale by animateFloatAsState(if (chosen) 1.12f else 1f, label = "mood-$v")
                    Column(
                        Modifier
                            .weight(1f)
                            .background(
                                if (chosen) Theme.primary.copy(alpha = 0.1f) else Color.Transparent,
                                RoundedCornerShape(12.dp),
                            )
                            .clickable(role = Role.RadioButton) { vibe = v }
                            .padding(vertical = 6.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(4.dp),
                    ) {
                        Text(
                            MoodScale.emoji(v),
                            Modifier
                                .scale(scale)
                                // Once a value is picked the others dim rather than vanish:
                                // the scale has to stay readable while you compare it.
                                .alpha(if (vibe == null || chosen) 1f else 0.4f),
                            fontSize = 28.sp,
                        )
                        Text(
                            MoodScale.label(v),
                            fontSize = 11.sp,
                            color = if (chosen) Theme.primary else Theme.inkSoft,
                            textAlign = TextAlign.Center,
                        )
                    }
                }
            }

            SectionLabel("Note (optional)")
            OutlinedTextField(
                value = note,
                onValueChange = { note = it },
                placeholder = { Text("What's on your mind?", color = Theme.inkSoft) },
                modifier = Modifier.fillMaxWidth().heightIn(min = 80.dp),
            )

            if (existing != null) {
                TextButton(onClick = { store.clearMood(dateIso); onDismiss() }) {
                    Text("Remove mood", color = Theme.phaseMenstrual)
                }
            }
        }
    }
}

// ------------------------------------------------------------------------------------
// Headache
// ------------------------------------------------------------------------------------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun HeadacheSheet(dateIso: String, existing: HeadacheLog?, onDismiss: () -> Unit) {
    val store = LocalCycleStore.current
    var intensity by remember { mutableIntStateOf(existing?.intensity ?: 2) }
    val symptoms = remember { existing?.symptoms.orEmpty().toMutableStateList() }
    val triggers = remember { existing?.triggers.orEmpty().toMutableStateList() }
    var note by remember { mutableStateOf(existing?.note ?: "") }

    val initialTime = remember(existing) {
        runCatching { LocalDateTime.parse(existing?.at).toLocalTime() }.getOrDefault(LocalTime.of(9, 0))
    }
    // Follow the phone's own 12/24-hour setting rather than forcing one.
    val use24Hour = DateFormat.is24HourFormat(LocalContext.current)
    val timeState = rememberTimePickerState(initialTime.hour, initialTime.minute, is24Hour = use24Hour)

    // Collapsed until asked for. TimeInput grabs focus the moment it composes, throwing the
    // numeric keyboard over the symptom and trigger chips on every open; the onset time is
    // usually near enough already, so it stays a tappable summary until it isn't.
    var editingTime by remember { mutableStateOf(false) }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = Theme.surface,
    ) {
        SheetFrame(
            title = if (existing == null) "Headache" else "Edit headache",
            saveEnabled = true,
            onCancel = onDismiss,
            onSave = {
                // Combine the browsed day with the chosen time.
                val day = runCatching { LocalDate.parse(dateIso) }.getOrDefault(LocalDate.now())
                val at = day.atTime(timeState.hour, timeState.minute)
                if (existing == null) {
                    store.addHeadache(intensity, symptoms.toList(), triggers.toList(), note.trim(), at)
                } else {
                    store.updateHeadache(existing.id, intensity, symptoms.toList(), triggers.toList(), note.trim(), at)
                }
                onDismiss()
            },
        ) {
            SectionLabel("Intensity")
            SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
                (1..3).forEach { v ->
                    SegmentedButton(
                        selected = intensity == v,
                        onClick = { intensity = v },
                        shape = SegmentedButtonDefaults.itemShape(v - 1, 3),
                        colors = SegmentedButtonDefaults.colors(
                            activeContainerColor = Theme.primary.copy(alpha = 0.15f),
                            activeContentColor = Theme.primary,
                            inactiveContentColor = Theme.inkSoft,
                        ),
                    ) { Text(HeadacheScale.label(v)) }
                }
            }

            SectionLabel("Time")
            Row(
                Modifier
                    .fillMaxWidth()
                    .background(Theme.background, RoundedCornerShape(12.dp))
                    .clickable { editingTime = true }
                    .padding(vertical = 12.dp, horizontal = 16.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("Onset", Modifier.weight(1f), fontSize = 16.sp, color = Theme.ink)
                Text(
                    formatTime(timeState.hour, timeState.minute, use24Hour),
                    Modifier
                        .background(Theme.surface, RoundedCornerShape(8.dp))
                        .padding(horizontal = 12.dp, vertical = 6.dp),
                    fontSize = 16.sp,
                    color = Theme.ink,
                )
            }

            SectionLabel("Symptoms")
            ChipWrap(SYMPTOM_OPTIONS, symptoms)

            SectionLabel("Possible triggers")
            ChipWrap(TRIGGER_OPTIONS, triggers)

            SectionLabel("Note (optional)")
            OutlinedTextField(
                value = note,
                onValueChange = { note = it },
                placeholder = { Text("Anything to remember", color = Theme.inkSoft) },
                modifier = Modifier.fillMaxWidth().heightIn(min = 72.dp),
            )

            if (existing != null) {
                TextButton(onClick = { store.deleteHeadache(existing.id); onDismiss() }) {
                    Text("Delete this episode", color = Theme.phaseMenstrual)
                }
            }
        }
    }

    // The dial in a dialog, rather than a text field inline: the inline numeric input takes
    // focus the moment it appears, so the keyboard covered the symptom and trigger chips and
    // the sheet had to be scrolled blind. The dial needs no keyboard at all.
    if (editingTime) {
        AlertDialog(
            onDismissRequest = { editingTime = false },
            containerColor = Theme.surface,
            title = { Text("Onset time", style = serif(20).copy(color = Theme.ink)) },
            text = {
                TimePicker(
                    state = timeState,
                    colors = TimePickerDefaults.colors(
                        selectorColor = Theme.primary,
                        containerColor = Theme.background,
                        periodSelectorSelectedContainerColor = Theme.primary.copy(alpha = 0.15f),
                        timeSelectorSelectedContainerColor = Theme.primary.copy(alpha = 0.15f),
                    ),
                )
            },
            confirmButton = {
                TextButton(onClick = { editingTime = false }) {
                    Text("Done", color = Theme.primary)
                }
            },
        )
    }
}

/**
 * A wrapping set of multi-select chips, two equal columns wide — the same tidy grid the iOS
 * sheet uses. Ragged widths made a list of nine triggers hard to scan.
 */
@Composable
private fun ChipWrap(options: List<String>, selection: MutableList<String>) {
    FlowRow(
        Modifier.fillMaxWidth().padding(vertical = 4.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
        maxItemsInEachRow = 2,
    ) {
        options.forEach { option ->
            val on = option in selection
            Text(
                option,
                Modifier
                    .weight(1f)
                    .background(if (on) Theme.primary else Theme.background, CircleShape)
                    .clickable(role = Role.Checkbox) {
                        if (on) selection.remove(option) else selection.add(option)
                    }
                    .padding(vertical = 10.dp, horizontal = 8.dp),
                fontSize = 13.sp,
                color = if (on) Color.White else Theme.ink,
                textAlign = TextAlign.Center,
            )
        }
    }
}

private fun formatTime(hour: Int, minute: Int, use24Hour: Boolean): String {
    if (use24Hour) return "%02d:%02d".format(hour, minute)
    val suffix = if (hour < 12) "AM" else "PM"
    val h = when {
        hour % 12 == 0 -> 12
        else -> hour % 12
    }
    return "%d:%02d %s".format(h, minute, suffix)
}

// ------------------------------------------------------------------------------------
// Note
// ------------------------------------------------------------------------------------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NoteSheet(dateIso: String, existing: JournalEntry?, onDismiss: () -> Unit) {
    val store = LocalCycleStore.current
    var text by remember { mutableStateOf(existing?.text ?: "") }
    val remaining = MAX_NOTE_LENGTH - text.length

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = Theme.surface,
    ) {
        SheetFrame(
            title = if (existing == null) "New note" else "Edit note",
            saveEnabled = true,
            onCancel = onDismiss,
            onSave = {
                val t = text.trim()
                if (t.isNotEmpty()) {
                    if (existing == null) store.addJournalEntry(t, dateIso)
                    else store.updateJournalEntry(existing.id, t, dateIso)
                }
                onDismiss()
            },
        ) {
            OutlinedTextField(
                value = text,
                onValueChange = { if (it.length <= MAX_NOTE_LENGTH) text = it },
                modifier = Modifier.fillMaxWidth().heightIn(min = 180.dp),
                textStyle = serif(16).copy(color = Theme.ink),
            )
            // Only surface the limit as it comes into view — a counter on an empty note is
            // just clutter.
            if (remaining <= 200) {
                Text(
                    "$remaining characters left",
                    Modifier.fillMaxWidth(),
                    fontSize = 11.sp,
                    color = if (remaining <= 0) Theme.phaseMenstrual else Theme.inkSoft,
                    textAlign = TextAlign.End,
                )
            }
            if (existing != null) {
                TextButton(onClick = { store.deleteJournalEntry(existing.id); onDismiss() }) {
                    Text("Delete", color = Theme.phaseMenstrual)
                }
            }
        }
    }
}

// ------------------------------------------------------------------------------------
// Shared sheet chrome
// ------------------------------------------------------------------------------------

@Composable
private fun SheetFrame(
    title: String,
    saveEnabled: Boolean,
    onCancel: () -> Unit,
    onSave: () -> Unit,
    content: @Composable () -> Unit,
) {
    Column(Modifier.fillMaxWidth()) {
        // The header stays put while the fields scroll under it, so Save is always reachable
        // — the headache sheet is taller than the sheet itself on a small phone.
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 20.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TextButton(onClick = onCancel) { Text("Cancel", color = Theme.inkSoft) }
            Text(
                title,
                Modifier.weight(1f),
                style = serif(20).copy(color = Theme.ink, textAlign = TextAlign.Center),
            )
            TextButton(onClick = onSave, enabled = saveEnabled) {
                Text(
                    "Save",
                    color = if (saveEnabled) Theme.primary else Theme.inkSoft.copy(alpha = 0.5f),
                    fontWeight = FontWeight.SemiBold,
                )
            }
        }
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp)
                .padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            content()
        }
    }
}

@Composable
private fun SectionLabel(text: String) {
    Text(
        text.uppercase(),
        Modifier.padding(top = 6.dp),
        fontSize = 11.sp,
        letterSpacing = 0.8.sp,
        color = Theme.accentText,
    )
}

/** "Headaches & your cycle" — the hormonal-cluster insight, only shown when confident. */
@Composable
fun HeadacheInsightCard(phaseLabel: String, count: Int, total: Int) {
    Row(
        Modifier.cyclunaCard(),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Box(Modifier.padding(top = 2.dp)) { HeartGlyph(Theme.secondary, 18.dp) }
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text("Headaches & your cycle", style = serif(19).copy(color = Theme.ink))
            Text(
                "Your headaches tend to cluster in your ${phaseWord(phaseLabel)}, " +
                    "a common hormonal pattern ($count of $total so far).",
                fontSize = 13.sp,
                color = Theme.inkSoft,
            )
        }
    }
}

/** Phase label → the phrase the sentences use. */
fun phaseWord(phaseLabel: String): String = when (phaseLabel) {
    "Ovulatory" -> "ovulation window"
    "Follicular" -> "follicular phase"
    "Luteal" -> "luteal phase"
    else -> "period"
}
